### ISSUE TITLE
update_index only when specific fields change?

### ISSUE THREAD

**Issue by eliduke** [toptal/chewy#134]
The User model in my project has a lot of fields that are updated regularly for certain metrics. It has `sign_in_count` and a couple other fields that get updated as a user is clicking around the site. As a result, the UsersIndex is constantly being updated because the user is being saved each time those fields get updated. It is possible update_index only when specific fields change?

I am imagining something like this:

`update_index('users', only: [:first_name, :last_name, :email]) { self }`

Thanks!

**Comment by pyromaniac** [comment]
Yeah, I was thinking about it, even attempting to derive such a list of fields automatically. but currently the best solution would be just like this: `update_index('users') { self if changes.include?(:first_name, :last_name, :email) }` if you are returning nil form the block - chewy will reindex nothing, so feel free to add any conditions you want.

**Comment by eliduke** [comment]
I'm getting a `wrong number of arguments (4 for 1)` error. All the documentation I'm seeing about `include?` shows that it can only receive 1 argument. Am I missing some brackets or something?

`update_index('users') { self if changes.include?(:first_name, :last_name, :organization_name, :title) }`

**Comment by eliduke** [comment]
This is what worked for me:

`update_index('users') { self if (changes.keys & [:first_name, :last_name, :title, :organization_name]).present? }`

**Comment by pyromaniac** [comment]
Yeah, sure, I was just wrong, `.include?` takes an array

**Comment by eliduke** [comment]
I spoke too soon!

That solution works in general, but I had to tweak a few things and then that got me wondering about best practices with `update_index`. Here's the deal.

I have users and users have many groups through user_groups. In my UsersIndex, a user has an array of group_ids. When a user joins another group, it doesn't actually update the user (it updates user_groups), so the UsersIndex doesn't get reindexed. But I DO want the user to be reindexed to include the new group in their group_ids. So, I added a callback method to UserGroup like this:

```
after_save :update_users_index
after_destroy :update_users_index

def update_users_index
  self.user.save!
end
```

That works pretty well, but because I specified that the user had to have changes that corresponded to specific fields, that callback approach didn't work anymore. So, I added to to my whitelist method like this:

```
update_index('users') { self if user_has_changes?}

def user_has_changes?
  changes.empty? || (changes.keys & ["first_name", "last_name", "title", "organization_name"]).present?
end
```

That works now. And normal clicking around doesn't update the user because the user is being updated regularly with little metrics. A user with absolutely no changes is unusual, accept during those callbacks.

But then I started thinking, is there a better way with `update_index`. Right now, I only have an update_index in the models that directly correspond to the indexes that I have. I have a User model and a UsersIndex and in the User model I have an `update_index`. Is it best practice to put other `update_index`s in other models?

Could I put an `update_index` in the UserGroup model that updates the user?

```
class UserGroup < ActiveRecord::Base

  update_index('users') {self.user}

end
```

My question is, when is that `update_index` called? Where does it fall in the `before` and `after` callbacks? Will it update the user index when a UserGroup is created and destroyed?

Thanks again for all your help!
